c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -1763,6 +2135,146 @@ class DeviceCachingAllocator {
     return blocks;
   }
 
+  // returns the smallest possible address in any segment
+  // where there is enough free address space to fit size
+  // may be composed of free and unmapped segments
+  Block* find_expandable_block(
+      int device,
+      cudaStream_t stream,
+      BlockPool* pool,
+      size_t size) {
+    Block key(device, stream, 0);
+
+    auto allocateable = [](Block* b) {
+      return b && !b->allocated && b->event_count == 0 &&
+          b->stream_uses.empty();
+    };
+    auto has_available_address_space = [&](Block* b) {
+      size_t bytes = 0;
+      while (bytes < size && allocateable(b)) {
+        bytes += b->size;
+        b = b->next;
```

Comment:
I'm surprised we would have adjacent, unallocated blocks. I thought that breaks an invariant

Reply:
allocatable blocks are always free but they might be unmapped or mapped. The adjacent ones are pairs of unmapped/mapped.